### PR TITLE
React/week4

### PR DESCRIPTION
**Pull request by Enqira**

### CODE REVIEW

File: react/week4/git-api/src/components/FormComponent.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  return (
+    <div>
+      <form>
+        <p>Search for git users:</p>
+        <input type="text" onChange={handleChange} placeholder="search" />
+      </form>
+      {users.length >= 1 ? (
+        <ul>
+          {!loading ? (
+            users.map(user => <li>{user.login}</li>)
```
Comment (calm-prog) [inline react/week4/git-api/src/components/FormComponent.js:20]: Don't forget to give your list items a key (which can be the index). This way React can keep better track of changes in the list. 
https://reactjs.org/docs/lists-and-keys.html

File: react/week4/git-api/src/components/UserContext.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  const [users, setUsers] = useState([])
+
+  useEffect(() => {
+    if (input) {
+      setLoading(true)
+      fetch(`${API}${input}`)
+        .then(res => res.json())
+        .then(data => {
+          data.message
+            ? alert("API rate limit exeeded, please wait some seconds")
```
Comment (calm-prog) [inline react/week4/git-api/src/components/UserContext.js:17]: An alternative solution to the rate limit issue is to fetch items only with the first character, then save that in a state, and when the user enters the next character, you just filter down the already fetched array of items. Then you will have to make a new fetch only when the user removes the word completely and starts writing a new word. 
This way you can reduce the number of fetches to a minimum level, and the user can search for items without alert message popping up.

File: react/week4/git-api/src/components/UserContext.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import React, { createContext, useEffect, useState } from "react"
+
+const API = `https://api.github.com/search/users?q=`
+
+export const UserContext = createContext()
+
+export const UserProvider = ({ children, input, loading, setLoading }) => {
```
Comment (calm-prog) [inline react/week4/git-api/src/components/UserContext.js:7]: It's great that you noticed that you can extract this logic into a separate component. Now the App component is much more transparent. Good thinking!

**Review by calm-prog — COMMENTED** [review]
Hey, I had a look at your pull request and had a couple of thoughts. 
You've completed the task well. Your code is well-structured and easy to read. You've extracted your global state logic into a separate component which makes it more transparent how states are managed.

However, not all your states are managed by the Context API. "Input" and "loading" are passed down as props, these could also be managed by the UserContext component. Arguably, the idea behind using Context in a small application like this is to keep all your state in a central "store" (while in a larger application you would probably opt for Redux). 

In addition to the requirements, there is something else that you might consider:
In this application you weren't in need of moving up the state to a central level. In fact, you could have handled all your states in the single Form component. 
The _current Form component_, however, fulfills 2 roles: it takes in a user input and outputs the search results. So you see these _could_ actually _be separated into 2 components_, Form and SearchResults. Since these 2 components would be interconnected, now you would be left with the only option of moving up the state and make use of Context as well.

**Comment by Enqira** [comment]
Thank you very much for your review. I will take that into considiration and  put into practice what you mentioned.

**Pull request merged**